src/dataFiles.ts

```diff
@@ -39,65 +36,104 @@ export async function initializeDataDirs(): Promise<void> {
 	});
 	if (!dataDirs?.length) return;
 	for (const dataDir of dataDirs) {
-		watchers.set(dataDir, null);
+		modifiedPaths.set(dataDir, new Set());
+		watchers.set(
+			dataDir,
+			watch(dataDir, { recursive: true, persistent: false }, (_, f) => {
+				if (!f) return;
+				const fullPath = resolve(join(dataDir, f));
+				if (fullPath === resolve(dataDir)) return;
+				modifiedPaths.get(dataDir)!.add(fullPath);
+			}),
+		);
 	}
 	logger.info("Indexing dataDirs for reverse lookup...");
-	await indexDataDirs();
+	await indexDataPaths(findSearcheesFromAllDataDirs());
+}
+
+/**
+ * This gets called on each rss/webhook/announce event.
+ */
+export async function indexDataDirs(): Promise<void> {
+	const { maxDataDepth } = getRuntimeConfig();
+	await Promise.all(
+		Array.from(watchers.keys()).map((dataDir) => {
+			const modified = modifiedPaths.get(dataDir)!;
+			const eventPaths: string[] = [];
+			while (modified.size) {
+				const path: string | undefined = modified.values().next().value;
+				if (!path) continue;
+				if (!modified.delete(path)) continue;
+				eventPaths.push(path);
+			}
+			if (!eventPaths.length) return;
+			logger.verbose(
+				`Indexing dataDir ${dataDir} due to recent changes...`,
+			);
+			eventPaths.sort(
+				(a, b) =>
+					b.split(sep).filter(isTruthy).length -
+					a.split(sep).filter(isTruthy).length,
+			);
+			const seenPaths = new Set<string>();
+			const paths = eventPaths.reduce<string[]>((acc, path) => {
+				const affectedPaths: string[] = [path];
+				let parentPath = dirname(path);
+				while (resolve(parentPath) !== resolve(dataDir)) {
+					affectedPaths.push(parentPath);
+					parentPath = dirname(parentPath);
+				}
+				for (const affectedPath of affectedPaths.slice(-maxDataDepth)) {
+					if (seenPaths.has(affectedPath)) continue;
+					seenPaths.add(affectedPath);
+					acc.push(affectedPath);
+				}
+				return acc;
+			}, []);
```

Comment:
seems like we could do this processing in multiple stages (flatmap all eventPaths into their ancestry chains, dedupe by throwing everything into a Set, then do the maxdatadepth check) and maybe have to juggle fewer variables at once.

Reply:
I'm not seeing the structure you're proposing. I want to handle the deepest paths first, as that makes the most of the memoization. Though it's not as important for reindexing since not much will change between rssCadence or announces. `eventPaths` won't have duplicates, but getting the parent paths will create duplicates. Having it grouped like this makes it easy to still obey `maxDataDepth`.